Fix SIGSEGV causes by unexpected cleanup of gangs

pengzhout opened the pull request:
There is a rule that when cleaning up gangs for a portal, it should not
clean up gangs belong to other portals, most code path follow the rule
except the cleanupPortalGangList(), this commit fixed this bug.

master has been refactored and has total different code path, so this is for 5x_stable only

pivotal-mike commented:
@pengzhout Does it make sense to port these test changes into Master as well?  If so can I ask you to open a new PR?

pengzhout commented:
Sure, will add those test cases to master too.

melanieplageman commented on src/test/regress/input/dispatch.source:
```diff
@@ -322,5 +322,30 @@ select gp_inject_fault('after_one_slice_dispatched', 'reset', 1);
+language plpgsql volatile;
+set gp_cached_segworkers_threshold to 1;
+select count(*) from foo, bar where exists (select foo_func());
+reset gp_cached_segworkers_threshold;
```
What is the relationship between this test and the code that is being fixed? What does setting gp_cached_segworkers_threshold to 1 do? When we ran this test on 5X_STABLE, we did notice that we tripped this assertion `FailedAssertion("!(b->slice->primaryGang->gang_id != 1)"` but we weren't sure 
1) Why this assertion was tripped in the unpatched code vs the patched code 
2) What the relationship between the GUC `gp_cached_segworkers_threshold` and this assertion is
It would be nice to clarify that in the comment above the test.

pengzhout replied:
I always hit a SIGSEGV in my local lab, the Assertion seems reasonable because the memory context of b->slice->primaryGang is freed, so the context is undetermined

pengzhout replied:
gp_cached_segworkers_threshold is a threshold to control the number of idle gangs.

melanieplageman commented on src/backend/cdb/dispatcher/cdbgang.c:
```diff
@@ -1409,8 +1409,9 @@ cleanupPortalGangList(List *gplist, int cachelimit)
 
-		if (nLeft > cachelimit ||
-			getGangMaxVmem(gang) > gp_vmem_protect_gang_cache_limit)
+		if (isTargetPortal(gang->portal_name, portal_name) &&
```
One thing I am wondering is how this problem was not encountered in more cases. Since`cleanupPortalGangList` cleans up all eligible gangs in `availableReaderGangsN`, wouldn't this problem have surfaced more since whenever multiple portals have gangs in that list, they could potentially be cleaned up incorrectly?

pengzhout replied:
I'm also thinking this problem should be exposed earlier or more frequently, one explanation is that the default value of gp_cached_segworkers_threshold is 5 and we need a init plan at the same time and a cursor is contained in init plan also.

pengzhout commented:
Modified the commit message and add more comments in test cases

pengzhout commented:
> Could you add to the commit message a description of the impetus for this fix (i.e. the bug that caused the crash)?

Added more comments and modified the commit message, please take another look.

pengzhout commented:
Merged, thanks for the review.
